jetty-util/src/main/java/org/eclipse/jetty/util/resource/ZipFsResource.java

```diff
@@ -0,0 +1,306 @@
+    {
+        this(file.toURI(), MR_MODE_RUNTIME);
+    }
+
+    public ZipFsResource(URI uri) throws IOException
+    {
+        this(uri, MR_MODE_RUNTIME);
+    }
+
+    public ZipFsResource(URI uri, String multiReleaseMode) throws IOException
+    {
+        zipFsUri = toZipFsUri(uri);
+        URI fileUri = URI.create(uri.getSchemeSpecificPart());
+        zipFsPath = new File(fileUri.getSchemeSpecificPart()).toPath();
+
+        // The FileSystem Environment
+        env = new HashMap<>();
+        env.put("multi-release", multiReleaseMode);
+
+        // The ZipFs FileSystem
+        FileSystem fs;
+        try
+        {
+            fs = FileSystems.getFileSystem(zipFsUri);
+        }
+        catch (FileSystemNotFoundException ignore)
+        {
+            fs = null;
+        }
+
+        if ((fs == null) || !fs.isOpen())
+        {
+            fs = FileSystems.newFileSystem(zipFsUri, env);
+        }
+
+        zipFs = fs;
+
+        // We only care about the 1 root that a zip/jar has
+        Iterable<Path> rootDirs = zipFs.getRootDirectories();
+        rootPath = rootDirs.iterator().next();
+    }
+
+    private final URI toZipFsUri(URI uri)
+    {
+        String asciiUri = uri.toASCIIString();
+
+        if (asciiUri.equalsIgnoreCase("jar:file:") &&
+                asciiUri.endsWith(".jar"))
+        {
+            // we have what we need already.
+            return uri;
+        }
+
+        if ("file".equalsIgnoreCase(uri.getScheme()))
+        {
+            if (asciiUri.endsWith(".jar"))
+            {
+                return URI.create("jar:" + asciiUri);
+            }
+        }
+
+        throw new UnsupportedOperationException(ZipFsResource.class.getSimpleName() + " only supports [jar:file:/] or [file:/] URIs that end with [.jar] - " + uri);
+    }
+
+    @Override
+    public boolean isContainedIn(Resource resource) throws MalformedURLException
+    {
+        if (resource instanceof PathResource)
+        {
+            PathResource pathResource = (PathResource) resource;
+            Path internalPath = zipFs.getPath(resource.getName());
+            try
+            {
+                return Files.isSameFile(pathResource.getPath(), internalPath);
+            }
+            catch (IOException ignore)
+            {
+                LOG.ignore(ignore);
+                return false;
+            }
+        }
+        // all other resource types are not supported here
+        return false;
+    }
+
+    @Override
+    public void close()
+    {
+        try
+        {
+            zipFs.close();
+        }
+        catch (IOException ignore)
+        {
+            LOG.ignore(ignore);
+        }
+    }
+
+    @Override
+    public boolean exists()
+    {
+        // If we passed the constructor, this is known to exist
+        return true;
+    }
+
+    @Override
+    public boolean isDirectory()
+    {
+        // a ZipFsResource is always a directory (never a file)
```

Comment:
really?   `jar:file:///some/path/file.zip!/other/path/index.html` is not a directory!

Reply:
yes. and `jar:file:///some/path/file.zip!/other/path/index.html` is a `java.nio.file.Path` pointing to the `java.nio.file.FileSystem` (re; [Path.getFileSystem()](https://docs.oracle.com/javase/8/docs/api/java/nio/file/Path.html#getFileSystem--)) referenced by `jar:file:///some/path/file.zip`

Think of it this way ...

    jar:file://<filesystem>!/<path-in-zipfs-filesystem>

Reply:
Oooooooooh! sound of light bulb turning on... or of penny dropping!!!

I had assumed that a Path vs File was kind of like URI vs URL, ie that the former was just the name itself unbound to any concrete provisioning of the underlying resource.    So a `Path` contains a file system!!!!!!! I thought it was that a `Path` was interpreted in the context of a file system!   

This makes more sense now.... and perhaps makes the handling of file systems much much easier and we might be able to get away with a very minimal caching layer... perhaps even none at all.

So I need to spend some time with this now to re-interpret all my other comments, as they will mostly be incorrect.